### PR TITLE
[OS-980, OS-1022] Threads: Add thread tests and fixes for KanoWorld

### PR DESCRIPTION
**Pull request by tombettany**
Add tests for multi-threaded access to the `KanoWorld` class and utilise
the thread sanitizer.

### CODE REVIEW

File: test/thread/TestKanoWorldRefresh.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/**
+ * \file kw_apis.h
```
Comment (radujipa) [inline test/thread/TestKanoWorldRefresh.h]:
```suggestion
 * \file TestKanoWorldRefresh.h
```

File: test/test.cpp [hunk 2]
```diff
@@ -25,6 +25,9 @@
 
 #include "theme/theme.h"
 
+// Run thread tests through the address sanitizer
+#include "thread/TestKanoWorldRefresh.h"
```
Comment (radujipa) [inline test/test.cpp]: Just a question: If we're making a separated multi-thread tests program, why are we including it in the unit ones here?

File: test/thread/TestKanoWorldRefresh.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+#include <gmock/gmock.h>
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#include <chrono>  // NOLINT
+#include <functional>
+#include <random>
+#include <string>
+#include <thread>  // NOLINT
```
Comment (radujipa) [inline test/thread/TestKanoWorldRefresh.h]: Interesting, linter complaining about `<thread>` also?

File: test/test.cpp (hunk 2 above)
Comment (skarbat) [reply test/test.cpp]: I second this question

File: test/thread/TestKanoWorldRefresh.h [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+
+TEST(KWRefreshThread, RefreshVerified) {
+    Mercury::KanoWorld::KanoWorld kw;
+    std::thread refresh_thr(
+        &Mercury::KanoWorld::KanoWorld::refresh_account_verified,
+        &kw,
+        false);
+
```
Comment (skarbat) [inline test/thread/TestKanoWorldRefresh.h]: should there be a swift pause here to let the thread breathe?

File: test/thread/TestKanoWorldRefresh.h [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+TEST(KWReadWriteThread, ReadAndWriteTokenFromThreads) {
+    int thread_count = 50;
+    std::string token = "testtoken";
+    Mercury::KanoWorld::KanoWorld kw;
+    std::vector<std::thread> threads;
+
+    for (int i = 0; i < thread_count; i++) {
+        std::thread write_thr([token, &kw]() {
+            random_delay(1000);
+            kw.set_token(token);
```
Comment (skarbat) [inline test/thread/TestKanoWorldRefresh.h]: should we assert that the token made it into the class correctly?

File: test/thread/TestKanoWorldRefresh.h (hunk 3 above)
Comment (tombettany) [reply test/thread/TestKanoWorldRefresh.h]: Because Chromium has implementations of things which overlap with `<thread>`, `<future>`, `<mutex>` and `<condition_variable>` (i.e. the C++ standard threading libraries) these are forbidden by `cpplint` (a tool made by Google). See [here](https://chromium-cpp.appspot.com/) for more info.

File: test/test.cpp (hunk 2 above)
Comment (tombettany) [reply test/test.cpp]: There are three sanitisers that we can use: address, undefined behaviour and thread. (There is also a memory one but I've not got that working yet). The thread and address sanitisers are mutually exclusive in their application so we need a separate program to allow us to run the thread sanitiser (the main tests run with the address sanitiser). The reason that we include the thread tests here is because we are just trying to catch errors (like memory frees, etc) encountered by the use in threads while the thread test binary catches data races.

Long story short, we run them in both to allow all the sanitisers to catch problems even though we don't necessarily assert.

File: test/thread/TestKanoWorldRefresh.h (hunk 5 above)
Comment (tombettany) [reply test/thread/TestKanoWorldRefresh.h]: Ideally we would put some asserts here to test that the results are what we expect but the main focus of the initial version of these tests is to simply ensure that it doesn't randomly fall over when we spin up threads.
